sys/include/net/gcoap.h

```diff
@@ -378,12 +381,14 @@ typedef struct {
  * @brief  Container for the state of gcoap itself
  */
 typedef struct {
+    mutex_t lock;                       /**< Shares state attributes safely */
     gcoap_listener_t *listeners;        /**< List of registered listeners */
     gcoap_request_memo_t open_reqs[GCOAP_REQ_WAITING_MAX];
                                         /**< Storage for open requests; if first
                                              byte of an entry is zero, the entry
                                              is available */
-    uint16_t last_message_id;           /**< Last message ID used */
+    atomic_uint_least16_t next_message_id;
```

Comment:
I didn't want to use atomic_uint_least16_t, but unfortunately, there is no atomic_uint16_t or atomic_uint16! (There is an atomic_uint and atomic_ushort). I went with the "least" variant because I wanted to be sure I had 2 bytes. That is the width of the value in the CoAP header spec and in coap_hdr_t.

I thought about just using atomic_uint, but that would mean casting with each use. That's kind of error prone to remember if we need to use this function again. I'd also probably want to create a local variable to temporarily store the narrowed value. Otherwise, the expression would be getting pretty long for a parameter to coap_build_hdr(). I don't feel strongly about using atomic_uint though. Let me know what you think.